### PR TITLE
LIBCLOUD-625: Allow for internal GCE authorization with metadata service

### PR DESCRIPTION
**Pull request by erjohnso**
This PR provides support for authorizing requests from within GCE directly by pulling an authorization token out of GCE's internal metadata services.  This greatly simplifies using libcloud inside GCE since it does not require the user to copy around private auth files.  The only parameter required for this method is the user's Project ID.  See the included docs for a sample use-case.

This PR also attempts to address https://issues.apache.org/jira/browse/LIBCLOUD-607 by appending the Project ID to the cached credential file.

### CODE REVIEW

**Comment by erjohnso** [comment]
/me sighs - I never seem to get Travis to smile on the first try...  These errors seem unrelated to the PR.

```
pypy runtests: commands[0] | cp libcloud/test/secrets.py-dist libcloud/test/secrets.py
WARNING:test command found but not installed in testenv
  cmd: /bin/cp
  env: /home/travis/build/apache/libcloud/.tox/pypy
Maybe forgot to specify a dependency?
```

File: libcloud/common/google.py [hunk 1]
```diff
@@ -99,6 +99,25 @@
...
+def _isGce():
+    http_code, http_reason, body = _getGceMetadata()
+    if http_code == httplib.OK and body:
+        return True
+    return False
+
+
+def _getGceMetadata(path='/computeMetadata/v1/'):
+    try:
+        md_http = httplib.HTTPConnection("metadata:80", timeout=2)
```
Comment (Kami) [inline libcloud/common/google.py]: Please use `libcloud.utils.connection.get_response_object` method. There is an example of how to use it at https://github.com/apache/libcloud/blob/trunk/libcloud/pricing.py#L203

This method uses Libcloud HTTP machinery underneath. Eventually, we will probably switch to the `requests` library, but now this is the way to go.

File: libcloud/common/google.py [hunk 2]
```diff
@@ -99,6 +99,25 @@
 TIMESTAMP_FORMAT = '%Y-%m-%dT%H:%M:%SZ'
 
 
+def _isGce():
```
Comment (Kami) [inline libcloud/common/google.py]: Please use underscore separated function names.

File: libcloud/compute/drivers/gce.py [hunk 3]
```diff
@@ -569,19 +569,30 @@ def __init__(self, user_id, key, datacenter=None, project=None,
...
+        :keyword  credential_file: Path to file for caching authentication
+                                   information used by GCEConnection.
+        :type     credential_file: ``str``
         """
+
         self.auth_type = auth_type
         self.project = project
         self.scopes = scopes
+        self.credential_file = credential_file or '~/.gce_libcloud_auth'
+        self.credential_file += '.' + self.project
```
Comment (Kami) [inline libcloud/compute/drivers/gce.py]: Hm, do we always want to append project to the credentials_file or only when the credential_file is not provided?

Comment (erjohnso) [inline libcloud/compute/drivers/gce.py]: Good point, I'll fix that.

File: libcloud/common/google.py (hunk 1 above)
Comment (erjohnso) [inline libcloud/common/google.py]: Works for me. The only issue I see is that Google's metadata service requires a custom request header.  I'll make a small change to `get_response_object` to support custom headers.

**Comment by erjohnso** [comment]
Hot diggity! Travis is even happy (after fixing my legitimate errors...)

**Comment by Kami** [comment]
@erjohnso Great, thanks.

I'm having problems cleanly applying patch to the trunk. Can you please sync your branch with trunk and squash the commits?

**Comment by erjohnso** [comment]
@Kami - can you give this a try?  I think it's safe to disregard the scary merge-conflict warning. I did manually resolve the conflicts on `libcloud/utils/connection.py` and `libcloud/compute/drivers/gce.py` and rebased.  I'm not sure what I did, but if you run into issues with this attempt, I'm going to whack this PR and create a new one.

**Comment by Kami** [comment]
@erjohnso 

```
kami /w/lc/libcloud (git:trunk)$ git am --signoff < 20d977075117f05a0d8cd8ceb91c4dfcd93a7766.patch
Applying: LIBCLOUD-625: Allow for internal GCE authorization with metadata service
error: patch failed: libcloud/compute/drivers/gce.py:550
error: libcloud/compute/drivers/gce.py: patch does not apply
error: patch failed: libcloud/utils/connection.py:21
error: libcloud/utils/connection.py: patch does not apply
Patch failed at 0001 LIBCLOUD-625: Allow for internal GCE authorization with metadata service
The copy of the patch that failed is found in:
   /w/lc/libcloud/.git/rebase-apply/patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort"
```

It's fine, I will try to pull down your branch and merge it directly into trunk.

**Comment by Kami** [comment]
Merged into trunk. Thanks.

There was a minor merge conflict - can you please make sure that everything still works as expected in trunk?

Thanks.

**Comment by erjohnso** [comment]
@Kami - thank you!  Tested out trunk and it's working as expected.

**Pull request merged**